hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/DumpReplicationQueues.java

```diff
@@ -294,80 +327,99 @@ public String dumpPeersState(List<ReplicationPeerDescription> peers) throws Exce
+        Collections.sort(wals);
+        if (!peerIds.contains(queueId.getPeerId())) {
+          deletedQueues.add(regionServer + "/" + queueId);
+          sb.append(formatQueue(regionServer, queueStorage, wals, queueId, true, hdfs));
+        } else {
+          sb.append(formatQueue(regionServer, queueStorage, wals, queueId, false, hdfs));
+        }
+      }
+    }
     return sb.toString();
   }
 
-  private String formatQueue(ServerName regionserver, ReplicationQueueStorage queueStorage,
-    ReplicationQueueInfo queueInfo, String queueId, List<String> wals, boolean isDeleted,
-    boolean hdfs) throws Exception {
+  private String formatQueue(ServerName regionServer, ReplicationQueueStorage queueStorage,
+    List<String> wals, ReplicationQueueId queueId, boolean isDeleted, boolean hdfs)
+    throws Exception {
     StringBuilder sb = new StringBuilder();
 
-    List<ServerName> deadServers;
-
-    sb.append("Dumping replication queue info for RegionServer: [" + regionserver + "]" + "\n");
+    sb.append("Dumping replication queue info for RegionServer: [" + regionServer + "]" + "\n");
     sb.append("    Queue znode: " + queueId + "\n");
-    sb.append("    PeerID: " + queueInfo.getPeerId() + "\n");
-    sb.append("    Recovered: " + queueInfo.isQueueRecovered() + "\n");
-    deadServers = queueInfo.getDeadRegionServers();
-    if (deadServers.isEmpty()) {
-      sb.append("    No dead RegionServers found in this queue." + "\n");
+    sb.append("    PeerID: " + queueId.getPeerId() + "\n");
+    sb.append("    Recovered: " + queueId.isRecovered() + "\n");
+    if (queueId.getSourceServerName().isPresent()) {
+      sb.append("    Dead RegionServers: " + queueId.getSourceServerName().get() + "\n");
     } else {
-      sb.append("    Dead RegionServers: " + deadServers + "\n");
+      sb.append("    No dead RegionServers found in this queue." + "\n");
     }
     sb.append("    Was deleted: " + isDeleted + "\n");
     sb.append("    Number of WALs in replication queue: " + wals.size() + "\n");
-    peersQueueSize.addAndGet(queueInfo.getPeerId(), wals.size());
+    peersQueueSize.addAndGet(queueId.getPeerId(), wals.size());
+
+    Set<Map.Entry<String, ReplicationGroupOffset>> offsets =
+      queueStorage.getOffsets(queueId).entrySet();
 
     for (String wal : wals) {
-      // long position = queueStorage.getWALPosition(regionserver, queueInfo.getPeerId(), wal);
-      // sb.append(" Replication position for " + wal + ": "
-      // + (position > 0 ? position : "0" + " (not started or nothing to replicate)") + "\n");
+      long position = -1;
+      String walGroup = null;
+      for (Map.Entry<String, ReplicationGroupOffset> entry : offsets) {
+        walGroup = entry.getKey();
+        ReplicationGroupOffset offset = entry.getValue();
+        if (offset.getWal().equals(wal)) {
+          position = offset.getOffset();
+        }
+      }
+      sb.append(" Replication position for " + (walGroup != null ? walGroup + "/" + wal : wal)
```

Comment:
Notice that, -1 in ReplicationGroupOffset means this file has already been fully replicated.